src/System.Linq.Expressions/src/System/Linq/Expressions/StackGuard.cs

```diff
@@ -0,0 +1,91 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Linq.Expressions
+{
+    internal sealed class StackGuard
+    {
+        private const int StackProbingThreshold = 100;
+        private const int MaxExecutionStackCount = 1024;
+
+        private int _recursionDepth;
+        private int _executionStackCount;
+
+        public bool TryEnterOnCurrentStack()
+        {
+            if (++_recursionDepth >= StackProbingThreshold)
+            {
+                try
+                {
+                    RuntimeHelpers.EnsureSufficientExecutionStack();
+                }
+                catch (InsufficientExecutionStackException) when (_executionStackCount < MaxExecutionStackCount)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        public void Exit()
+        {
+            _recursionDepth--;
+        }
+
+        public void RunOnEmptyStack(Action<object> action, object state)
+        {
```

Comment:
Actually, it occurs to me.  Since all of the call sites are taking tuples, if you were willing to have customized overloads for each number of arguments, you could do something along the lines of:
```C#
    static void Run<T1, T2>(Action<T1, T2> action, T1 arg1, T2 arg2)
    {
        Task.Factory.StartNew(s =>
        {
            var t = (Tuple<Action<T1, T2>, T1, T2>)s;
            t.Item1(t.Item2, t.Item3);
        }, Tuple.Create(action, arg1, arg2));
    }
```
the idea being that you move the tuple creation to inside of the method, thereby allowing you to include the provided action in the tuple as well, providing a delegate to StartNew that extracts the real delegate to use.